src/aspire/source/simulation.py

```diff
@@ -468,3 +475,132 @@ def eval_coords(self, mean_vol, eig_vols, coords_est):
+            sim.rotations = sim.rotations[np.random.permutation(sim.n)]
+
+        # Assert NoiseAdder has not been provided
+        if "noise_adder" in kwargs or sim.noise_adder is not None:
+            raise RuntimeError(
+                "Cannot provide 'noise_adder' when using {cls.__name__}.from_snr."
+            )
+
+        # Estimate the required noise variance
+        signal_var = sim.estimate_signal_power(
+            sample_n=sample_n,
+            support_radius=support_radius,
+            signal_power_method=signal_power_method,
+        )
+        noise_var = signal_var / target_snr
+
+        # Assign the noise_adder
+        sim.noise_adder = WhiteNoiseAdder(var=noise_var)
+        logger.info(f"Appended {sim.noise_adder} to generation pipeline")
+
+        # Restore the original rotations when previously shuffled.
+        if shuffle:
+            sim.rotations = _rots
+
+        return sim
+
+    def estimate_psnr(self, sample_n=None, batch_size=512, units=None):
+        """
+        Estimate Peak SNR as max(signal)^2 / MSE,
+        where MSE is computed between `projections`
+        and the resulting simulated `images`.
+        PSNR is computed along the stack axis and an average
+        value across the stack sample is returned.
+        Note that PSNR is inherently a poor metric for identical images.
+
+        :param sample_n: Number of images used for estimate.
+            Defaults to all images in source.
+        :param units: Optionally, convert from default ratio to log scale (`dB`).
+        :returns: Estimated peak signal to noise ratio.
+        """
+
+        if sample_n is None:
+            sample_n = self.n
+
+        if sample_n > self.n:
+            logger.warning(
+                f"`estimate_psnr` sample_n > Source.n: {sample_n} > {self.n}."
+                f" Accuracy may be impaired, settting sample_n=self.n={self.n}"
+            )
+            sample_n = self.n
+
+        peaksq = np.empty(sample_n, dtype=self.dtype)
+        mse = np.empty(sample_n, dtype=self.dtype)
+        for start in trange(0, sample_n, batch_size):
+            end = min(start + batch_size, sample_n)
+
+            signal = self.projections[start:end].asnumpy()
+            images = self.images[start:end].asnumpy()
+            peaksq[start:end] = np.square(signal).max(axis=(-1, -2))  # max per image
+
+            # Reshape and Transpose for Scikit metrics,
+            # which expect a 2d (samples, outputs)
+            mse[start:end] = mean_squared_error(
+                signal.reshape(sample_n, -1).T,
+                images.reshape(sample_n, -1).T,
+                multioutput="raw_values",
+            )
+
+        psnr = peaksq / mse
+        if units == "dB":
+            psnr = ratio_to_decibel(psnr)
+        elif units is not None:
+            raise ValueError("Units should be `None` or `dB`.")
+
+        # Return the mean psnr of the stack.
+        return np.mean(psnr)
```

Comment:
There's a number of ambiguities here: average before or after division (i.e. peak square value over all the images divided by MSE over all the images vs. what we have here) but also mean before or after log (corresponds to arithmetic vs. geometric mean). I'm not familiar enough with PSNR to say which is the more common convention.